sycl/doc/extensions/SubGroup/SYCL_INTEL_sub_group.asciidoc

```diff
@@ -0,0 +1,259 @@
+  {
+    /* kernel code */
+  }
+}
+----
+
+It is illegal for a kernel or function to call a function with a mismatched sub-group size requirement, and the compiler should produce an error in this case.  The +reqd_sub_group_size+ attribute is not propagated from a device function to callers of the function, and must be specified explicitly when a kernel is defined.
+
+=== Sub-group Queries
+
+Several aspects of sub-group functionality are implementation-defined: the size and number of sub-groups is implementation-defined (and may differ for each kernel); and different devices may make different guarantees with respect to how sub-groups within a work-group are scheduled.  Developers can query these behaviors at a device level and for individual kernels.  The sub-group size for a given combination of kernel and launch configuration is fixed, and guaranteed to be reflected by device and kernel queries.
+
+Each sub-group in a work-group is one-dimensional.  If the total number of work-items in a work-group is evenly divisible by the sub-group size, all sub-groups in the work-group will contain the same number of work-items.  If the total number of work-items in a work-group is not evenly divisible by the sub-group size, the number of work-items in the final sub-group is equal to the remainder of the total work-group size divided by the sub-group size.
+
+To maximize portability across devices, developers should not assume that work-items within a sub-group execute in lockstep, nor that two sub-groups within a work-group will make independent forward progress with respect to one another.
+
+The device descriptors below are added to the +info::device+ enumeration class:
+
+|===
+|Device Descriptors|Return Type|Description
+
+|+info::device::max_num_sub_groups+
+|+uint32_t+
+|Returns the maximum number of sub-groups in a work-group for any kernel executed on the device.  The minimum value is 1.
+
+|+info::device::sub_group_independent_forward_progress+
+|+bool+
+|Returns +true+ if the device supports independent forward progress of sub-groups with respect to other sub-groups in the same work-group.
+
+|+info::device::sub_group_sizes+
+|+vector_class<size_t>+
+|Returns a vector_class of +size_t+ containing the set of sub-group sizes supported by the device.
+|===
+
+An additional query for sub-group information is added to the +kernel+ class:
+
+|===
+|Member Functions|Description
+
+|+template <info::kernel_sub_group param>typename info::param_traits<info::kernel_sub_group, param>::return_type get_sub_group_info(const device &dev) const+
+|Query information from the sub-group from a kernel using the +info::kernel_sub_group+ descriptor for a specific device.
+|===
+
+The kernel descriptors below are added as part of a new +info::kernel_sub_group+ enumeration class:
+
+|===
+|Kernel Descriptors|Return Type|Description
+
+|+info::kernel_sub_group::max_num_sub_groups+
+|+uint32_t+
+|Returns the maximum number of sub-groups for this kernel.
```

Comment:
@Pennycook, 

I've just now realized that the list of info descriptors was reduced and previously we had:
1. `info::kernel_sub_group::local_size_for_sub_group_count` - removed now
2. `info::kernel_sub_group::sub_group_count_for_ndrange` - removed now
3.  `info::kernel_sub_group::max_sub_group_size_for_ndrange` - replaced with `info::kernel_sub_group::max_sub_group_size`

I agree that (1) is not really needed and can be replaced by the following sequence:
- ensure that desired number of sub-groups doesn't exceed `max_num_sub_groups`
- query sub-group size selected for the kernel
- multiple desired number of sub-groups by selected sub-group size
- ensure that calculated value doesn't exceed maximum work-group size for a kernel
- if previous check passed, use calculated work-group size

(2) is even easier to replace - once you know selected sub-group size, you can easily calculate amount of sub-groups per work-group

So, (1) and (2) depend on selected sub-groups size for a kernel, which is intended to be returned by (3) and here is a problem: according to OpenCL 2.1 spec, rev. 23:

> Work items are mapped into sub-groups through a combination of compile-time decisions **\*and
the parameters of the dispatch\***. The mapping to sub-groups is invariant for the duration of a
kernel’s execution, **\*across dispatches of a given kernel with the same work-group dimensions\***,
between dispatches and query operations consistent with the dispatch parameterization, and from
one work-group to another within the dispatch (excluding the trailing edge work-groups in the
presence of non-uniform work-group sizes).

Spec explicitly states that selected sub-groups size for a particular kernel might be changed from one dispatch to another if work-group size is changed.

Strictly speaking, we cannot query exact `max_sub_group_size` from OpenCL RT without specifying exact work-group size which will be used for kernel dispatch.

Reply:
Sorry for the delayed response.  I've opened https://github.com/intel/llvm/pull/1565 to address this.